Polishing pass on Tock 2.0 system call TRD. (#2431)
tock/tock: doc/reference/trd104-syscalls.md

Every system call MUST return only one failure and only one success
variant. Different system calls may use different failure and success
variants, but any specific system call returns exactly one of each. If an
operation might have multiple success return variants or failure return
variants, then it MUST be split into multiple system calls.

This requirement of a single failure variant and a single success
variant is to simplify userspace implementations and preclude them
from having to handle many different cases.  The presence of many
difference cases suggests that the operation should be split up, as
there is non-determinism in its execution or its meaning is

> overloaded. The requirement of a single failure anda single success

Reviewer: ```suggestion
overloaded. The requirement of a single failure and a single success
```

| 1024  | BADRVAL     | The variant of the return value did not match what the system call should return.       |

Values in the range 1-1023 reflect kernel return value error
codes. Kernel error codes not specified above are reserved. TRDs MAY
specify additional kernel error codes from these reserved values, but
MUST NOT specify kernel error codes greater than 1023.  The Tock
kernel MUST NOT return an error code unless the error code is
specified in a TRD.

Values greater than 1023 are reserved for userspace library use. Value 1024
(BADRVAL) is for when a system call returns a different failure or success
variant than the userspace library expects. 

4 System Call API
=================================

Tock has 7 classes or types of system calls. When a system call is
invoked, the class is encoded as the Syscall Class ID. Some system
call classes are implemented by the core kernel and so the supported
calls are the same across kernels. Others are implemented by system
call drivers, which can be added and removed in different kernel
builds. The full set of valid system calls a kernel supports therefore
depends on what system call drivers it has installed.

> The 6 classes are:

Reviewer: Right above the table it says "The 6 classes are:" when there are 7 classes.

Author: Fixed

| Memop            |        5         |
| Exit             |        6         |

All of the system call classes except Yield and Exit are
non-blocking. When a userspace process calls a Subscribe, Command,
Allow, Read-Only Allow, or Memop syscall, the kernel will not put it
on a wait queue. Instead, it will return immediately upon
completion. The kernel scheduler may decide to suspend the process due
to a timeslice expiration or the kernel thread being runnable, but the
system calls themselves do not block. If an operation is long-running
(e.g., I/O), its completion is signaled by an upcall (see the

> Subscribe call in 4.2).

Reviewer 2: I would have written this that system calls themselves _do_ block. For example, every memop blocks the process from running until the entire memop operation has completed. Only then will the kernel reschedule the process.

Also, with the new scheduler trait, there is no guarantee that a process will not be put on a wait queue after calling a syscall.

Author: That's not traditionally what blocking means. Non-blocking I/O calls, for example, still return a return value. Usually blocking means that calling the function can cause the process to be descheduled and put on a wait queue, or forced to wait in some way.

Reviewer 3: I agree with @phil-levis  that "blocking" implies blocking on I/O (or some other long operation).

@bradjc is right that there is no type- or implementation- guarantee, but this _is_ the semantics of system calls, and it _should_ be implemented by any scheduler (unless of course that scheduler is exploring ways in which different semantics might be better). In other words, it is the semantic that a Tock kernel promises to processes.

Successful calls to Exit system calls do not return (the process exits).

System calls implemented by system call drivers (Subscribe, Command,
Allow, Read-Only Allow) all include two arguments, a driver identifier
and a syscall identifier. The driver identifier specifies which system
call driver to invoke. The syscall identifier (which is different than
the Syscall Class ID in the table above) specifies which instance of
that system call on that driver to invoke. Both arguments are unsigned
32-bit integers. For example, the Console system call driver has

> driver identifier `0x1` and a Command to the console driver with

Reviewer 2: ```suggestion
driver identifier `0x1` by default and a Command to the console driver with
```

Author: You can change the mappings? This seems extremely dubious. If you can change the mappings then we shouldn't have standard ones/a registry.

Reviewer 3: I think this is a persistent confusion that ought to be clarified, but probably not here. There are several notions of what "Tock" is, at least: a kernel crate that provides a particular execution model, interfaces for capsules, low-level system call semantics (the 2.0 system call interface defined in this document), etc; also an operating system implementation that uses that kernel and provides certain basic services (a console, timers, etc), with particular standardized system call interfaces.

In the latter case, the Console driver has driver identifier `0x1`. In former case, there may or may not be a console driver at all, it might be at a different system call number, and may have a totally different interface. Maybe it's used to store beanie baby futures.

I think this document is defining the latter---the low level system call interface that the kernel enforces, not the standard set of services provided via the system call interface to boards by convention.

Also, maybe this particular use of language doesn't really matter here, as it's just an example, and the example is correct. Would using "by convention" rather than "by default" be OK with everyone (which I think is technically more precise, though at the risk of being unnecessarily verbose)?

| Memop            |        5         |
| Exit             |        6         |

> All of the system call classes except Yield and Exit are
> non-blocking. When a userspace process calls a Subscribe, Command,
> Allow, Read-Only Allow, or Memop syscall, the kernel will not put it
> on a wait queue. Instead, it will return immediately upon
> completion. The kernel scheduler may decide to suspend the process due
> to a timeslice expiration or the kernel thread being runnable, but the
> system calls themselves do not block. If an operation is long-running
> (e.g., I/O), its completion is signaled by an upcall (see the
> Subscribe call in 4.2).

Reviewer 2: ```suggestion
All of the system call classes except Yield and Exit are
non-blocking. When a userspace process calls a Subscribe, Command,
Allow, Read-Only Allow, or Memop syscall, the kernel will not put the process
on a wait queue while handling the syscall. Instead, the kernel will complete the
syscall and prepare the return value for the syscall immediately.
The kernel scheduler may not, however, run the process immediately after
handling the syscall, and may instead decide to suspend the process due
to a timeslice expiration or the kernel thread being runnable.
If an operation is long-running
(e.g., I/O), its completion is signaled by an upcall (see the
Subscribe call in 4.2).
```

Successful calls to Exit system calls do not return (the process exits).

System calls implemented by system call drivers (Subscribe, Command,
Allow, Read-Only Allow) all include two arguments, a driver identifier
and a syscall identifier. The driver identifier specifies which system
call driver to invoke. The syscall identifier (which is different than
the Syscall Class ID in the table above) specifies which instance of
that system call on that driver to invoke. Both arguments are unsigned
32-bit integers. For example, the Console system call driver has

> driver identifier `0x1` and a Command to the console driver with

Reviewer 2: ```suggestion
driver identifier `0x1`, by convention, and a Command to the console driver with
```

| Memop            |        5         |
| Exit             |        6         |

All of the system call classes except Yield and Exit are
non-blocking. When a userspace process calls a Subscribe, Command,

> Allow, Read-Only Allow, or Memop syscall, the kernel will not put the process

Reviewer: ```suggestion
Read-Write Allow, Read-Only Allow, or Memop syscall, the kernel will not put the process
```

on a wait queue while handling the syscall. Instead, the kernel will complete the
syscall and prepare the return value for the syscall immediately.
The kernel scheduler may not, however, run the process immediately after
handling the syscall, and may instead decide to suspend the process due
to a timeslice expiration or the kernel thread being runnable.
If an operation is long-running
(e.g., I/O), its completion is signaled by an upcall (see the
Subscribe call in 4.2).

Successful calls to Exit system calls do not return (the process exits).

System calls implemented by system call drivers (Subscribe, Command,

> Allow, Read-Only Allow) all include two arguments, a driver identifier

Reviewer: ```suggestion
Read-Write Allow, Read-Only Allow) all include two arguments, a driver identifier
```

Tock Binary Format for application images mean that the first address
will not be executable code and so 0x0 will not be a valid function.
In the case that 0x0 is valid application code and where the
linker places an upcall function, the first instruction of the function
should be a no-op and the address of the second instruction passed
instead.

If a userspace process invokes subscribe on a driver ID that is not
installed in the kernel, the kernel MUST return a failure with an

> error code of NOSUPPORT and an upcall of the Null Upcall.

Reviewer 2: ```suggestion
error code of `NODEVICE` and an upcall of the Null Upcall.
```

Per line 272? Although per the code it's `NOSUPPORT`.

Author: Good catch.

its failure variant and success variant. If userspace invokes a
command on a peripheral that is not installed, the kernel returns a
failure variant of `Failure`, with an associated error code of

> `NOSUPPORT`. Therefore, command invocations that need to

Reviewer 2: There are a few references to `NOSUPPORT`, including the current implementation:

https://github.com/tock/tock/blob/38c12029d76530847591aded6afe14008f49668e/kernel/src/driver.rs#L186-L193

Author: Yeah, I think there's a discrepancy between the code and the error values. The kernel should return `ErrorCode::NODEVICE` if there is no `Driver` implementation for the device ID, but a system call driver should return 'ErrorCode::NOSUPPORT` if it's an invalid system call ID.

At least, this is what the ErrorCode/ReturnCode values seem to say to me, unless we want to interpret a "device" as something else. Or we could just tweak this text and come up with a definition of NODEVICE that's distinct from NOSUPPORT and UNINSTALLED.

```

The `success` field indicates whether the call succeeded.
If it failed, the error code is stored in `error`. If it succeeded,
the value in `error` is undefined. `ptr` and `size` contain the pointer
and size of the passed buffer.

5.1.6 Memop
---------------------------------

Because the Memop system calls are defined by the kernel and not extensible, they are
directly defined by libtock-c as library functions:

```c
void* tock_app_memory_begins_at(void);
void* tock_app_memory_ends_at(void);
void* tock_app_flash_begins_at(void);
void* tock_app_flash_ends_at(void);
void* tock_app_grant_begins_at(void);
int tock_app_number_writeable_flash_regions(void);
void* tock_app_writeable_flash_region_begins_at(int region_index);
void* tock_app_writeable_flash_region_ends_at(int region_index);
```

They wrap around an underlying function which uses inline assembly:

```c
void* memop(uint32_t op_type, int arg1);
```

5.1.7 Exit
---------------------------------

The Exit system calls have these function prototypes:

```c
void tock_exit(uint32_t completion_code);
void tock_restart(uint32_t completion_code);
```

Since these two variants of Exit never return, they have
no return value.

> 5.2 libtock-rs

Reviewer 4: `libtock-rs` will have a more complex syscalls API than `libtock-c`, as there will be multiple variations of Subscribe, Read-Only Allow, and Read-Write Allow with different lifetime semantics. I also expect it to keep changing for a while.

Should we get rid of the `libtock-rs` heading?

Author: Sure -- the goal here was to show examples. Having a follow-up TRD that specifies libtock-rs would be fine.
